src/poetry/puzzle/provider.py

```diff
@@ -856,6 +858,42 @@ def _is_relevant_marker(
             and (not self._env or marker.validate(self._env.marker_env))
         )
 
+    def _add_implicit_dependencies(
+        self,
+        dependencies: Iterable[Dependency],
+        active_extras: Collection[NormalizedName] | None,
+    ) -> list[Dependency]:
+        """
+        This handles an edge case where the dependency is not required for a subset
+        of possible environments. We have to create such an implicit "not required"
+        dependency in order to not miss other dependencies later.
+        For instance:
+            • foo (1.0) ; python == 3.7
+            • foo (2.0) ; python == 3.8
+            • bar (2.0) ; python == 3.8
+            • bar (3.0) ; python == 3.9
+        the last dependency would be missed without this,
+        because the intersection with both foo dependencies is empty.
+
+        A special case of this edge case is a restricted dependency with a single
+        constraint, see https://github.com/python-poetry/poetry/issues/5506
+        for details.
+        """
+        by_name: dict[str, list[Dependency]] = defaultdict(list)
+        for dep in dependencies:
+            by_name[dep.name].append(dep)
+        for _name, deps in by_name.items():
+            marker = marker_union(*[d.marker for d in deps])
+            if marker.is_any():
+                continue
```

Comment:
Is `marker_union` also needed when e.g. `len(deps) == 1`? 
Because, at a glance, `marker_union` looks like a rather expensive function call.